src/util/arith_tools.h

```diff
@@ -11,21 +11,170 @@ Author: Daniel Kroening, kroening@kroening.com
+      "Numeric cast only works for types smaller than long long");
+#else
+    PRECONDITION(
+      std::numeric_limits<T>::max() <=
+        std::numeric_limits<decltype(mpi.to_long())>::max() &&
+      std::numeric_limits<T>::min() >=
+        std::numeric_limits<decltype(mpi.to_long())>::min());
+#endif
+    if(
+      mpi <= std::numeric_limits<T>::max() &&
+      mpi >= std::numeric_limits<T>::min())
+      // to_long converts to long long which is the largest signed numeric type
+      return static_cast<T>(mpi.to_long());
+    else
+      return {};
+  }
+
+  optionalt<T> operator()(const exprt &expr) const
+  {
+    auto mpi_opt = numeric_castt<mp_integer>{}(expr);
+    if(mpi_opt)
+      return numeric_castt<T>{}(*mpi_opt);
+    else
+      return {};
+  }
+};
+
+/// Convert mp_integer to any unsigned type
+/// \tparam T: type to convert to
+/// \param mpi: mp_integer to convert
+/// \return optional integer of type T if conversion is possible,
+///         empty optional otherwise.
+template <typename T>
+struct numeric_castt<T,
+  typename std::enable_if<std::is_integral<T>::value &&
+                          !std::is_signed<T>::value>::type>
+{
+  optionalt<T> operator()(const mp_integer &mpi) const
+  {
+#if !defined(_MSC_VER) || _MSC_VER >= 1900
+    static_assert(
+      std::numeric_limits<T>::max() <=
+      std::numeric_limits<decltype(mpi.to_ulong())>::max() &&
+      std::numeric_limits<T>::min() >=
+      std::numeric_limits<decltype(mpi.to_ulong())>::min(),
+      "Numeric cast only works for types smaller than unsigned long long");
+#else
+    // std::numeric_limits<> methods are not declared constexpr in old versions
+    // of VS
+    PRECONDITION(
+      std::numeric_limits<T>::max() <=
+        std::numeric_limits<decltype(mpi.to_ulong())>::max() &&
+      std::numeric_limits<T>::min() >=
+        std::numeric_limits<decltype(mpi.to_ulong())>::min());
+#endif
+    if(mpi <= std::numeric_limits<T>::max() && mpi >= 0)
+      return static_cast<T>(mpi.to_ulong());
+    else
+      return {};
+  }
+
+  optionalt<T> operator()(const exprt &expr) const
+  {
+    auto mpi_opt = numeric_castt<mp_integer>{}(expr);
+    if(mpi_opt)
+      return numeric_castt<T>{}(*mpi_opt);
+    else
+      return {};
+  }
+};
+
+/// Converts an expression to any integral type
+/// \tparam T: type to convert to
+/// \param arg: expression to convert
+/// \return optional integer of type T if conversion is possible,
+///         empty optional otherwise.
+template <typename Target>
+optionalt<Target> numeric_cast(const exprt &arg)
+{
+  return numeric_castt<Target>{}(arg);
+}
+
+/// Convert an mp_integer to integral type Target
+/// An invariant with fail with message "Bad conversion" if conversion
+/// is not possible.
+/// \tparam Target: type to convert to
+/// \param arg: mp_integer
+/// \return value of type Target
+template <typename Target>
+Target numeric_cast_v(const mp_integer &arg)
```

Comment:
What does `_v` mean?

Reply:
It stands for value (this is to distinguish from optionalt). This was the notation used in the `expr_cast` I removed, but I'm open to better suggestions.

Reply:
`_val`?